lgc/util/CacheHashNoteUtils.cpp

```diff
@@ -0,0 +1,298 @@
+using namespace lgc;
+
+// The implementation of ELF rewriting is based on "Linux Programmer's Manual ELF(5)".
+// In particular, see "Notes (Nhdr)" of the document for the note section.
+
+namespace {
+
+// It is similar to struct NoteHeader defined in llpc/util/vkgcElfReader.h, but
+// it allows us to use Elf_Nhdr_Impl<object::ELF64LE>::Align and it does not
+// have the size limitation of note name.
+using NoteHeader = object::Elf_Nhdr_Impl<object::ELF64LE>;
+
+// =====================================================================================================================
+// A section to shift and its new offset.
+struct SectionShiftInfo {
+  SmallString<64> section;
+  ELF::Elf64_Off newOffset;
+};
+
+// =====================================================================================================================
+// Create a note section with the note "llpc_cache_hash" that has the cache hash
+// as the description.
+//
+// Reference: Linux Programmer's Manual ELF(5) "Notes (Nhdr)".
+//
+// @param cacheHash               : Hash code to be added to the new .note section.
+// @param llpcVersion             : LLPC version that gives us a hint how cacheHash
+//                                was generated.
+// @param [out] noteSectionStream : The new .note section with the cache hash.
+void createNoteSectionWithCacheHash(StringRef cacheHash, StringRef llpcVersion,
+                                    AppendingBinaryByteStream &noteSectionStream) {
+  BinaryStreamWriter noteSectionWriter(noteSectionStream);
+
+  // Write the note header for the cache hash.
+  NoteHeader noteHeader;
+  noteHeader.n_namesz = sizeof(CacheHashNoteName);
+  noteHeader.n_descsz = cacheHash.size();
+  // TODO: Define a note type to specify cache hash. It must be defined in
+  // llvm/include/llvm/BinaryFormat/ELF.h.
+  // Note types with values between 0 and 32 (inclusive) are reserved.
+  noteHeader.n_type = 0;
+  auto error =
+      noteSectionWriter.writeFixedString(StringRef(reinterpret_cast<const char *>(&noteHeader), sizeof(noteHeader)));
+  assert(!error);
+
+  // Write the cache hash note name terminated by zero and zeros for the alignment.
+  error = noteSectionWriter.writeFixedString(StringRef(CacheHashNoteName, sizeof(CacheHashNoteName)));
+  assert(!error);
+  error = noteSectionWriter.writeFixedString(
+      StringRef("\0\0\0", offsetToAlignment(noteSectionWriter.getLength(), Align::Constant<NoteHeader::Align>())));
+  assert(!error);
+
+  // Write the cache hash and zeros for the alignment.
+  error = noteSectionWriter.writeFixedString(cacheHash);
+  assert(!error);
+  error = noteSectionWriter.writeFixedString(
+      StringRef("\0\0\0", offsetToAlignment(noteSectionWriter.getLength(), Align::Constant<NoteHeader::Align>())));
+  assert(!error);
+
+  // Write the note header for the llpc version note. The llpc version note can
+  // be used to determine the hash algorithm.
+  noteHeader.n_namesz = sizeof(LlpcVersionNoteName);
```

Comment:
To avoid repetition, would it be possible to write a function `addNoteEntry(StringRef noteName, stringRef noteDesc)` that does all of this?  Then you can call it twice.  Once for the hash and once for the version.

Reply:
done